File: PythonSolutions/Neetcode150/graph_problems/133_clone-graph.py

# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        if not node:
            return None
        
        oldToNew = {}

        def dfs(curr):
            # Base case: If node is already mapped, return the copied node
            if curr in oldToNew:
                return oldToNew[curr]
            
            copyNode = Node(curr.val) # Copy current node
            oldToNew[curr] = copyNode # Map current node to new copy

            # Go through all of neighbours for current node
            for nei in curr.neighbors:
                # Add the current nodes neighbors to the copied node
                copyNode.neighbors.append(dfs(nei))
            
            # Need to return copy node in the event base case is not hit for first time visits
            return copyNode
            
        return dfs(node)

File: TypescriptSolutions/Neetcode150/graph_grid_problems/133_clone-graph.ts

class _Node {
    val: number
    neighbors: _Node[]

    constructor(val?: number, neighbors?: _Node[]) {
        this.val = (val===undefined ? 0 : val)
        this.neighbors = (neighbors===undefined ? [] : neighbors)
    }
}



type GraphNode = _Node | null;

const cloneGraph = (node: GraphNode): GraphNode => {
    if (!node) return null;

    const oldToNew = new Map<GraphNode, GraphNode>();

    const dfs = (curr: GraphNode): GraphNode => {
        if (oldToNew.has(curr)) return oldToNew.get(curr)!;

        const copyNode = new _Node(curr!.val);
        oldToNew.set(curr, copyNode);

        for (const nei of curr!.neighbors) {
            copyNode.neighbors.push(dfs(nei)!);
        }

        return copyNode;
    };

    return dfs(node);
};